src/ERC6538Registry.sol

```diff
@@ -1,45 +1,125 @@
+  /// secp256k1, as specified in ERC-5564.
+  /// @param stealthMetaAddress The stealth meta-address to register.
   function registerKeys(uint256 schemeId, bytes memory stealthMetaAddress) external {
-    bytes memory registrant = _toBytes(msg.sender);
-    stealthMetaAddressOf[registrant][schemeId] = stealthMetaAddress;
-    emit StealthMetaAddressSet(registrant, schemeId, stealthMetaAddress);
+    stealthMetaAddressOf[msg.sender][schemeId] = stealthMetaAddress;
+    emit StealthMetaAddressSet(msg.sender, schemeId, stealthMetaAddress);
   }
 
-  /// @inheritdoc IERC6538Registry
+  /// @notice Sets the `registrant`'s stealth meta-address for the given scheme ID.
+  /// @param registrant Address of the registrant.
+  /// @param schemeId Identifier corresponding to the applied stealth address scheme, e.g. 0 for
+  /// secp256k1, as specified in ERC-5564.
+  /// @param signature A signature from the `registrant` authorizing the registration.
+  /// @param stealthMetaAddress The stealth meta-address to register.
+  /// @dev Supports both EOA signatures and EIP-1271 signatures.
+  /// @dev Reverts if the signature is invalid.
   function registerKeysOnBehalf(
     address registrant,
     uint256 schemeId,
     bytes memory signature,
-    bytes memory stealthMetaAddress
-  ) external pure {
-    registerKeysOnBehalf(_toBytes(registrant), schemeId, signature, stealthMetaAddress);
-  }
+    bytes memory stealthMetaAddress,
+    uint8 v,
+    bytes32 r,
+    bytes32 s
+  ) external {
+    bytes32 dataHash;
+    address recoveredAddress;
 
-  /// @inheritdoc IERC6538Registry
-  function registerKeysOnBehalf(
-    bytes memory, // registrant
-    uint256, // schemeId
-    bytes memory, // signature
-    bytes memory // stealthMetaAddress
-  ) public pure {
-    revert("not implemented");
+    unchecked {
+      dataHash = keccak256(
+        abi.encodePacked(
+          "\x19\x01",
+          domainSeparator,
+          keccak256(
+            abi.encode(TYPE_HASH, registrant, schemeId, stealthMetaAddress, nonceOf[registrant]++)
+          )
+        )
+      );
+      recoveredAddress = ecrecover(dataHash, v, r, s);
+    }
+
+    require(
+      (
+        (recoveredAddress != address(0) && recoveredAddress == registrant)
+          || (
+            IERC1271(registrant).isValidSignature(dataHash, signature)
+              == IERC1271.isValidSignature.selector
+          )
+      ),
+      "Invalid signature"
+    );
+
+    stealthMetaAddressOf[registrant][schemeId] = stealthMetaAddress;
+    emit StealthMetaAddressSet(registrant, schemeId, stealthMetaAddress);
   }
 
-  /// @dev Converts an `address` to `bytes`.
-  function _toBytes(address who) internal pure returns (bytes memory) {
-    return bytes.concat(bytes32(uint256(uint160(who))));
+  /// @notice Increments the nonce of the sender to invalidate existing signatures.
+  function incrementNonce() external {
```

Comment:
help me, when would this function be executed, which scenario?

Reply:
It was a helper function, but we don't need it anymore. Removed it here: [2384caa](https://github.com/ScopeLift/stealth-address-erc-contracts/pull/5/commits/2384caa12e103105097e644bdecf678b52918c3c)

Reply:
`incrementNonce` is needed to allow signers to revoke a signature that has not yet been submitted by incrementing their nonce, so it should not have been removed

Reply:
That's right, good catch. Added back in: [075b13d](https://github.com/ScopeLift/stealth-address-erc-contracts/pull/5/commits/075b13d1438122e6fe78af48224146f3d41c3cda)

Reply:
@mds1, thanks! this makes sense!